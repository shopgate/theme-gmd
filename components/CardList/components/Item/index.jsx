/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import ListItem from '@shopgate/pwa-common/components/List/components/Item';
import styles from './style';

/**
 * The Card List Item component implemented as class so that ref prop is available.
 */
class Item extends Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    isSelected: PropTypes.bool,
  };

  static defaultProps = {
    children: null,
    className: null,
    isSelected: false,
  };

  /**
   * Renders a card list item.
   * @returns {JSX|null}
   */
  render() {
    if (!Children.count(this.props.children)) {
      return null;
    }

    return (
      <ListItem className={`${styles} ${this.props.className}`} isSelected={this.props.isSelected}>
        {this.props.children}
      </ListItem>
    );
  }
}

export default Item;
