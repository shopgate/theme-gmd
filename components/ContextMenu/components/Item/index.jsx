/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import compose from 'lodash/fp/compose';
import Glow from 'Components/Glow';
import styles from './style';

/**
 * A delay in ms after that the closeMenu callback gets triggered.
 * @type {number}
 */
const CLOSE_DELAY = 250;

/**
 * The Context Menu Item component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Item = ({ children, closeMenu, onClick }) => {
  const handleClick = compose(
    onClick,
    () => setTimeout(closeMenu, CLOSE_DELAY)
  );

  return (
    <Glow>
      <div className={styles.item} onClick={handleClick} aria-hidden>
        {children}
      </div>
    </Glow>
  );
};

Item.propTypes = {
  children: PropTypes.node,
  closeMenu: PropTypes.func,
  onClick: PropTypes.func,
};

Item.defaultProps = {
  children: null,
  closeMenu: () => {},
  onClick: () => {},
};

export default Item;
