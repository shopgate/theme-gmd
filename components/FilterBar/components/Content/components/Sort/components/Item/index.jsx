/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Ripple from 'Components/Ripple';
import styles from './style';

/**
 * Renders the Sort component.
 * @param {Object} props The components props.
 * @param {Function} props.handleSelectionUpdate Callback that is executed when selection changed.
 * @returns {JSX}
 */
const Item = ({ children }) => (
  <Ripple fill className={styles}>
    {children}
  </Ripple>
);

Item.propTypes = {
  children: PropTypes.node,
};

Item.defaultProps = {
  children: null,
};

export default Item;
