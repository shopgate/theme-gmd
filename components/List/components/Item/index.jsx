/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Grid from '@shopgate/pwa-common/components/Grid';
import Link from '@shopgate/pwa-common/components/Router/components/Link';
import Glow from 'Components/Glow';
import styles from './style';

/**
 * The list item component.
 */
class Item extends Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
    className: PropTypes.string,
    image: PropTypes.element,
    isDisabled: PropTypes.bool,
    isSelected: PropTypes.bool,
    link: PropTypes.string,
    onClick: PropTypes.func,
    rightComponent: PropTypes.element,
  };

  static defaultProps = {
    className: null,
    image: null,
    isDisabled: false,
    isSelected: false,
    link: null,
    onClick: null,
    rightComponent: null,
  };

  /**
   * Should only update what the `selected` or `disabled` props change.
   * @param {Object} nextProps The next set of component props.
   * @returns {JSX}
   */
  shouldComponentUpdate(nextProps) {
    return (
      this.props.isSelected !== nextProps.isSelected ||
      this.props.isDisabled !== nextProps.isDisabled
    );
  }

  /**
   * Renders the bulk of the content.
   * @returns {JSX}
   */
  renderContent() {
    const { isDisabled, isSelected, title } = this.props;

    let gridStyles = styles.grid;
    let titleStyles = styles.title;

    if (isSelected) {
      gridStyles += ` ${styles.selected}`;
    }

    if (isDisabled) {
      titleStyles += ` ${styles.disabled}`;
    }

    return (
      <Grid className={gridStyles} component="div">
        <div className={styles.image}>
          {this.props.image}
        </div>
        <Grid.Item
          className={titleStyles}
          component="div"
          grow={1}
        >
          {title}
        </Grid.Item>
        {this.props.rightComponent &&
          <Grid.Item component="div" grow={1}>
            {this.props.rightComponent}
          </Grid.Item>
        }
      </Grid>
    );
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    /**
     * If this item is disabled, selected or doesn't have a valid
     * link or click handler then wrap the content with other components.
     */
    if (
      this.props.isDisabled ||
      (!this.props.link && !this.props.onClick)
    ) {
      return this.renderContent();
    }

    // Wrap with a <Link> if the `link` prop is set.
    if (this.props.link) {
      return (
        <Glow className={this.props.className}>
          <Link href={this.props.link} onClick={this.props.onClick}>
            {this.renderContent()}
          </Link>
        </Glow>
      );
    }

    return (
      <div aria-hidden onClick={this.props.onClick}>
        <Glow className={this.props.className}>
          {this.renderContent()}
        </Glow>
      </div>
    );
  }
}

export default Item;
