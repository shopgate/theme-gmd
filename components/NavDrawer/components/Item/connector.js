import { connect } from 'react-redux';
import { historyPush } from '@shopgate/pwa-common/actions/router';

/**
 * Maps action dispatchers to the component props.
 * @param {function} dispatch The store dispatcher.
 * @return {Object} The extended component props.
 */
const mapDispatchToProps = dispatch => ({
  navigate: (pathname, state) => dispatch(historyPush({
    pathname,
    state,
  })),
});

/**
 * @returns {boolean}
 */
const areStatePropsEqual = () => true;

export default connect(null, mapDispatchToProps, null, { areStatePropsEqual });
