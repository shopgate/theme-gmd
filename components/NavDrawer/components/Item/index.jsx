/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { history } from '@shopgate/pwa-common/helpers/router';
import ParsedLink from '@shopgate/pwa-common/components/Router/helpers/parsed-link';
import Grid from '@shopgate/pwa-common/components/Grid';
import Ripple from 'Components/Ripple';
import styles from './style';

// Add some delay to the click event to show the ripple effect.
const CLICK_DELAY = 250;

/**
 * Handles an Item click by executing it's href.
 * @param {Object} props The component props.
 * @param {string} props.href A url string.
 * @param {Function} props.close A callback.
 */
const handleClick = ({ onClick, href, close }) => {
  setTimeout(() => {
    // Perform onClick callback
    onClick();

    if (href) {
      // Open parsed link
      new ParsedLink(href).open(history);
    }

    // Call close callback from drawer
    close();
  }, CLICK_DELAY);
};
/**
 * The Item component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Item = (props) => {
  const className = classNames(
    styles.container, {
      [styles.primary]: props.primary,
    }
  );

  const labelClassName =
          props.withIndicator && !props.count ? styles.labelWithIndicator : styles.label;

  return (
    <div
      aria-hidden
      className={className}
      data-test-id="NavDrawerLink"
      onClick={() => handleClick(props)}
    >
      <Ripple fill>
        <Grid className={styles.grid}>
          <Grid.Item>
            <div className={styles.icon}>
              {props.icon && React.createElement(
                props.icon,
                {
                  ...props.primary && { className: styles.primaryIcon },
                }
              )}
            </div>
          </Grid.Item>
          <Grid.Item grow={1}>
            <div className={labelClassName}>
              {props.children}
            </div>
          </Grid.Item>
          {props.count && (
            <Grid.Item>
              <div className={styles.count}>
                {props.count}
              </div>
            </Grid.Item>
          )}
        </Grid>
      </Ripple>
    </div>
  );
};

Item.propTypes = {
  children: PropTypes.node,
  close: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  count: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  href: PropTypes.string, // eslint-disable-line react/no-unused-prop-types
  icon: PropTypes.func,
  onClick: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  primary: PropTypes.bool,
  withIndicator: PropTypes.bool,
};

Item.defaultProps = {
  children: null,
  close: () => {},
  count: null,
  href: '',
  icon: null,
  onClick: () => {},
  primary: false,
  withIndicator: false,
};

export default Item;
