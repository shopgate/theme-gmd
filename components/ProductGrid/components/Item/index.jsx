/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Grid from '@shopgate/pwa-common/components/Grid';
import Link from '@shopgate/pwa-common/components/Router/components/Link';
import Portal from '@shopgate/pwa-common/components/Portal';
import Ellipsis from '@shopgate/pwa-common/components/Ellipsis';
import { bin2hex } from '@shopgate/pwa-common/helpers/data';
import { PRODUCT_ITEM_NAME_BEFORE } from '@shopgate/pwa-common-commerce/product/constants/Portals';
import ProductImage from 'Components/ProductImage';
import DiscountBadge from 'Components/DiscountBadge';
import Price from 'Components/Price';
import PriceStriked from 'Components/PriceStriked';
import PriceInfo from 'Components/PriceInfo';
import FavoritesButton from 'Components/FavoritesButton';
import styles from './style';
import connect from './connector';

/**
 * The Product Grid Item component.
 * @param {Object} props The component props.
 * @return {JSX}
 */
const Item = ({ product, display, isFavorite }) => (
  <Link
    tagName="a"
    href={`/item/${bin2hex(product.id)}`}
    className={styles.container}
    itemProp="item"
    itemScope
    itemType="http://schema.org/Product"
  >
    <ProductImage itemProp="image" src={product.featuredImageUrl} alt={product.name} />
    {!!product.price.discount &&
      <div className={styles.badgeWrapper}>
        <DiscountBadge text={`-${product.price.discount}%`} />
      </div>
    }
    <div className={styles.favorites}>
      <FavoritesButton active={isFavorite} productId={product.id} noShadow removeWithRelatives />
    </div>
    {(!display || display.name || display.price || display.reviews) && (
      <div className={styles.details}>
        <Portal name={PRODUCT_ITEM_NAME_BEFORE} props={{ id: product.id }} />
        {(!display || display.name) && (
          <div className={styles.title} itemProp="name">
            <Ellipsis>{product.name}</Ellipsis>
          </div>
        )}
        {(!display || display.price) && (
          <Grid className={styles.priceWrapper} wrap>
            <Grid.Item grow={1}>
              <Price
                unitPrice={product.price.unitPrice}
                unitPriceMin={product.price.unitPriceMin}
                discounted={!!product.price.discount}
                currency={product.price.currency}
              />
            </Grid.Item>
            {product.price.msrp > 0 && (
              <Grid.Item>
                <PriceStriked
                  value={product.price.msrp}
                  currency={product.price.currency}
                />
              </Grid.Item>
            )}
            {(!product.price.msrp && product.price.unitPriceStriked > 0) && (
              <Grid.Item>
                <PriceStriked
                  value={product.price.unitPriceStriked}
                  currency={product.price.currency}
                />
              </Grid.Item>
            )}
          </Grid>
        )}
        {(!display || display.price) && (
          <Grid>
            {product.price.info && (
              <Grid.Item>
                <PriceInfo className={styles.basicPrice} text={product.price.info} />
              </Grid.Item>
            )}
          </Grid>
        )}
      </div>
    )}
  </Link>
);

Item.propTypes = {
  product: PropTypes.shape().isRequired,
  display: PropTypes.shape(),
  isFavorite: PropTypes.bool,
};

Item.defaultProps = {
  display: null,
  isFavorite: false,
};

export default connect(Item);
