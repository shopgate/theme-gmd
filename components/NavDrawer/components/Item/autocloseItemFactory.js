import React from 'react';
import Item from './index';

/**
 * Factory that creates an Item component which implements default close property.
 * Close property can still be overridden.
 * @param {function} close NavDrawer close function.
 * @returns {function}
 */
const autocloseItemFactory = (close) => (props) => (<Item close={close} {...props} />);

export default autocloseItemFactory;
