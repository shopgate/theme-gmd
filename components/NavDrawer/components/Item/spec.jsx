import React from 'react';
import { mount } from 'enzyme';
import Item from './index';
import autocloseItemFactory from './autocloseItemFactory';

describe('AutocloseItemFactory', () => {
  it('should create Item component with default close prop', () => {
     const closeFn = () => {};
     const AutocloseItem = autocloseItemFactory(closeFn);
     expect(typeof AutocloseItem === 'function');
     const component = mount(<AutocloseItem />);
     expect(component).toMatchSnapshot();
     expect(component.find('Item').props().close).toBe(closeFn);
     expect(component.find(Item).length).toBe(1);
  });
});