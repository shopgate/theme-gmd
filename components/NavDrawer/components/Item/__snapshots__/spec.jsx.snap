// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocloseItemFactory should create Item component with default close prop 1`] = `
<Component>
  <Item
    close={[Function]}
    count={null}
    href=""
    icon={null}
    link=""
    onClick={[Function]}
    primary={false}
    withIndicator={false}
  >
    <div
      aria-hidden={true}
      className="css-13azwyo"
      data-test-id="NavDrawerLink"
      onClick={[Function]}
    >
      <Ripple
        className=""
        color="#000"
        disabled={false}
        fill={true}
        onComplete={[Function]}
        overflow={false}
        size={null}
      >
        <div
          aria-hidden={true}
          className=""
          data-test-id="Ripple"
          onClick={[Function]}
          style={null}
        >
          <Grid
            className="css-1yzks37"
            component="ul"
            wrap={false}
          >
            <ul
              className="css-1yzks37 css-h7r55a"
            >
              <GridItem
                className=""
                component="li"
                grow={0}
                shrink={1}
              >
                <li
                  className=""
                >
                  <div
                    className="css-17ij4h5"
                  />
                </li>
              </GridItem>
              <GridItem
                className=""
                component="li"
                grow={1}
                shrink={1}
              >
                <li
                  className=" css-1gs0ko2"
                >
                  <div
                    className="css-1hv14w9"
                  />
                </li>
              </GridItem>
            </ul>
          </Grid>
        </div>
      </Ripple>
    </div>
  </Item>
</Component>
`;
